cmd/ipfs/daemon.go

```diff
@@ -463,6 +475,152 @@ func daemonFunc(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment
+func pinMFSOnChange(cctx *oldcmds.Context, node *core.IpfsNode) (<-chan error, error) {
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+		lastPins := map[string]lastPin{}
+		var tmo *time.Timer
+		defer func() {
+			if tmo != nil {
+				tmo.Stop()
+			}
+		}()
+		for {
+			// reread the config, which may have changed in the meantime
+			cfg, err := cctx.GetConfig()
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-cctx.Context().Done():
+					return
+				}
+			}
+			// if MFSRepinInterval is not set, pinning MFS is disabled,
+			// however this loop still polls to capture changes in the config
+			if cfg.Pinning.MFSRepinInterval == "" {
+				if tmo == nil {
+					tmo = time.NewTimer(time.Minute) // poll for config changes once per minute?
+				} else {
+					tmo.Reset(time.Minute)
+				}
+				select {
+				case <-cctx.Context().Done():
+					return
+				case <-tmo.C:
+				}
+				continue
+			}
+			mfsRepinInterval, err := time.ParseDuration(cfg.Pinning.MFSRepinInterval)
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-cctx.Context().Done():
+					return
+				}
+			}
+
+			// poll periodically
+			if tmo == nil {
+				tmo = time.NewTimer(mfsRepinInterval)
+			} else {
+				tmo.Reset(mfsRepinInterval)
+			}
+			select {
+			case <-cctx.Context().Done():
+				return
+			case <-tmo.C:
+			}
+			log.Infof("Pinning loop is awake")
+
+			// get the most recent MFS root cid
+			rootNode, err := node.FilesRoot.GetDirectory().GetNode()
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-cctx.Context().Done():
+					return
+				}
+				continue
+			}
+			rootCid := rootNode.Cid()
+
+			// pin on all remote services in parallel to prevent DoS attacks
+			ch := make(chan lastPin, len(cfg.Pinning.RemoteServices))
+			for svcName_, svcConfig_ := range cfg.Pinning.RemoteServices {
+				svcName, svcConfig := svcName_, svcConfig_
+				if svcConfig.Policies.PinMFS == nil || !*svcConfig.Policies.PinMFS {
+					continue
+				}
+				if last, ok := lastPins[svcName]; ok && last.ServiceConfig == svcConfig && last.CID == rootCid {
+					// do nothing, if MFS has not changed since last pin on the exact same service
+					ch <- lastPin{}
+					continue
+				}
+				log.Infof("Pinning MFS root %s to %s", rootCid, svcName)
+				go func() {
+					if r, err := pinMFS(cctx.Context(), node, rootCid, svcName, svcConfig, errCh); err != nil {
+						ch <- lastPin{}
+					} else {
+						ch <- r
+					}
+				}()
+			}
+			for i := 0; i < len(cfg.Pinning.RemoteServices); i++ {
+				x := <-ch
+				lastPins[x.ServiceName] = x
+			}
+		}
+	}()
+	return errCh, nil
+}
+
+func pinMFS(
+	ctx context.Context,
+	node *core.IpfsNode,
+	cid cid.Cid,
+	svcName string,
+	svcConfig config.RemotePinningService,
+	errCh chan<- error,
+) (lastPin, error) {
+	c := pinclient.NewClient(svcConfig.Api.ApiEndpoint, svcConfig.Api.ApiKey)
+
+	// Prepare Pin.name
+	opts := []pinclient.AddOption{pinclient.PinOpts.WithName("policy-mfs")}
```

Comment:
1.  Fixed. I missed it (and also forgot to update the go-ipfs-config version) when rebasing on master after merging the other PR :see_no_evil:

2. Not unreasonable to add the PeerID, although when the user does an `ipfs key rotate` the PeerID will change which might be a little annoying for users and also is something we'd need to consider if we poll the pinning services on startup at all.

3. I think having a "special" or "policy" marker in the name could be useful. I'm not normally a fan of the lexicographical race to the top, but this is probably fine.
   - We may want to consider the ordering of the name elements since we don't support regex type queries over the pins. e.g. are people more likely to want `__policy-mfs*` for all their peers or `__policy-peerid*` for all the policies for a given peer (assuming `peerid` is even relevant for other policies). It doesn't necessarily have to be the same ordering for MFS as for any other policy, but also worth a bit of considering about how this format might generalize to other policies.

Reply:
2. Another option is to let people set the name of the MFS pin in the policy itself. I tend to like this better than the peer-id (which can change).

Reply:
2. I would not worry about key rotation – no data is lost, it will rarely happen, and user can unpin old ones manually. 
   But I like Petar's idea: if we have `Policies.MFS` as struct, then `Policies.MFS.PinName` gives an  option to define deterministic pin name (when empty, default generated from peerid is used). 
3. Good point on "scoping" them per peerid. 
   With `__policy/peerid/mfs`, when sorted by name, pins produced by policies will be grouped per peerid (better UX)